File: Day-9/lru-cache-sample/java/LRUCache.java
import java.util.LinkedHashMap;
import java.util.Map;

public class LRUCache<K, V> {
    private final int capacity;
    private final LinkedHashMap<K, V> items;

    public LRUCache(int capacity) {
        if (capacity <= 0) {
            throw new IllegalArgumentException("capacity must be positive");
        }
        this.capacity = capacity;
        this.items = new LinkedHashMap<>(16, 0.75f, true);
    }

    public V get(K key) {
        return items.get(key);
    }

    public K put(K key, V value) {
        items.put(key, value);
        if (items.size() > capacity) {
            K evictedKey = items.keySet().iterator().next();
            items.remove(evictedKey);
            return evictedKey;
        }
        return null;
    }

    public Map<K, V> snapshot() {
        return new LinkedHashMap<>(items);
    }
}

File: Day-9/lru-cache-sample/python/lru_cache.py
from __future__ import annotations

from collections import OrderedDict
from dataclasses import dataclass, field
from typing import Generic, TypeVar

K = TypeVar("K")
V = TypeVar("V")


@dataclass
class LRUCache(Generic[K, V]):
    capacity: int
    _items: OrderedDict[K, V] = field(default_factory=OrderedDict, init=False)

    def __post_init__(self) -> None:
        if self.capacity <= 0:
            raise ValueError("capacity must be positive")

    def get(self, key: K) -> V | None:
        if key not in self._items:
            return None

        self._items.move_to_end(key)
        return self._items[key]

    def put(self, key: K, value: V) -> str | None:
        evicted: str | None = None
        if key in self._items:
            self._items.move_to_end(key)
        self._items[key] = value

        if len(self._items) > self.capacity:
            evicted_key, _ = self._items.popitem(last=False)
            evicted = str(evicted_key)
        return evicted

    def snapshot(self) -> list[tuple[K, V]]:
        return list(self._items.items())
